portfolio/src/main/java/com/google/sps/servlets/CommentsController.java

```diff
@@ -14,36 +14,239 @@
+      String comment = (String) entity.getProperty("comment");
+      long createdAt = (long) entity.getProperty("createdAt");
+      int likes = ((Long) entity.getProperty("likes")).intValue();
+
+      commentsList.add(new Comment(id, username, comment, createdAt, likes));
     }
+
+    int count = commentsList.size();
+    int pages = (int) Math.ceil((double) count / pageSize);
+    Integer next = page < pages ? (page + 1) : null;
+    Integer prev = page > 1 ? (page - 1) : null;
+
+    Info info = new Info(next, prev, count, pages);
+    int lowerIndex = (page - 1) * pageSize;
+    int upperIndex = page * pageSize;
+    List<Comment> trimmedComments = commentsList.subList(lowerIndex >= 0 ? lowerIndex : 0,
+        upperIndex <= count ? upperIndex : count);
+
+    Comments comments = new Comments(info, trimmedComments);
+    Gson gson = new Gson();
+    String data = gson.toJson(comments);
+
+    response.getWriter().println(data);
   }
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json;");
 
+    String body = getBody(request);
+    JSONObject json = new JSONObject();
+    try {
+      JSONParser parser = new JSONParser();
+      json = (JSONObject) parser.parse(body);
+    } catch (ParseException e) {
+      e.printStackTrace();
+    }
+
+    String username = getAttribute(json, "username", "Anonymous");
+    String comment = getAttribute(json, "comment", "...");
+    long createdAt = System.currentTimeMillis();
+    int likes = 0;
+
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("username", username);
+    commentEntity.setProperty("comment", comment);
+    commentEntity.setProperty("createdAt", createdAt);
+    commentEntity.setProperty("likes", likes);
+
+    datastore.put(commentEntity);
+
+    long id = commentEntity.getKey().getId();
+
     Gson gson = new Gson();
-    String data = gson.toJson(comments);
+    String data = gson.toJson(new Comment(id, username, comment, createdAt, likes));
+
+    response.getWriter().println(data);
+  }
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json;");
+
+    String body = getBody(request);
+    JSONObject json = new JSONObject();
+
+    try {
+      JSONParser parser = new JSONParser();
+      json = (JSONObject) parser.parse(body);
+    } catch (ParseException e) {
+      e.printStackTrace();
+    }
+
+    String id = getAttribute(json, "id", null);
+    Key key = KeyFactory.createKey("Comment", Long.parseLong(id));
+
+    Entity commentEntity;
+    try {
+      commentEntity = datastore.get(key);
+    } catch (EntityNotFoundException e) {
+      e.printStackTrace();
+      return;
```

Comment:
I was thinking on handling errors by sending a JSON with the corresponding HTTP status code, `400` for this particular case. But I think that it would've taken too much work and it's not part of the walkthrough.

By default, the server returns `500` when something brokes on any request, but it sends an HTML template that can't be parsed by the `.json()` used in the client. I think that's enough as long as I connect correctly the client.

If you think that error handling in JSON format should be added both here and the other try/catch blocks, please let me know.

Reply:
I think this is fine as well - thanks for bringing it up!